Nebulaworks/orion
# [READY] [infra - 532] Term-apply DynamoDB backend

## Description

Fixes: [infra #532](https://gitlab.com/nebulaworks/eng/infra/-/issues/532)

Changes data storage for term-apply from csv files stored in S3 to a dedicated DynamoDB table that can easily interface with future applicant review software.

The expected schema for each item is as follows:
> applied_date: unix time - number - Sort DDB Key
> email: candy@date.com - string - Primary/Partition DDB Key
> name: Candy Date - string
> github: candydate100 - string - Secondary Global Index
> role_applied: sr. software engineer - string
> role_override: string - in the case were their role is different otherwise null
> resume_review: bool - Resume passed or fail the review
> resume_review_date: number - resume reviewed date
> interviews: Map
>   1: {date: number, pass: bool, notes: binary}
>   2: {date: number, pass: bool, notes: binary}
> ignore_workflow: bool - ignore from automation workflow
> offer_given: bool
> offer_date: number
> offer_accepted:  bool
> offer_accepted_date: number
> rejected: bool
> rejected_date: number
> rejected_msg_override: binary - Message custom to applicant 

However term-apply only interfaces with a subset of attributes:
> applied_date: unix time - number - Sort DDB Key
> email: candy@date.com - string - Primary/Partition DDB Key
> github: candydate100 - string - Secondary Global Index
> name: Candy Date - string
> role_applied: sr. software engineer - string
> offer_given: bool
> rejected: bool

term-apply users have the ability to modify their email after submitting an application. If this happens, the existing record will be deleted and recreated with updated fields using a transaction to ensure atomic changes. Other item edits are done in place.

This system supports repeated applications by checking if a user's previous application has been closed (either `offer_given` or `rejected` is set to true). If a user has a closed application, a new one is created. Users can only modify their most recent open application.

## Tests
* Submitted application as a new user in empty table and populated table
  * new record successfully created in DynamoDB with correct fields in both cases
* Updated role for open application
  * Role successfully updated
* Updated name for open application
  * Name successfully updated
* Email updated for open application
  * delete and recreate transaction completed successfully
* Email updated for open application with other attributes populated
  * delete and recreate transaction completed successfully, no fields lost
* Role and Name updated for open with other attributes populated
  * Role and Name updated successfully with no other attributes changed
* Tests ran with multiple open applications
  * No side effects observed in other table items
* Application closed by setting `rejected` to true, then submitting a repeat application
  * User was able to reapply and a new record was created in DynamoDB
* Tests ran with repeated application
  * Only the most recent open application was modified, as expected
